sql/core/src/main/scala/org/apache/spark/sql/execution/exchange/ShuffleExchangeExec.scala

```diff
@@ -275,8 +282,17 @@ object ShuffleExchangeExec {
           position
         }
       case h: HashPartitioning =>
-        val projection = UnsafeProjection.create(h.partitionIdExpression :: Nil, outputAttributes)
-        row => projection(row).getInt(0)
+        val extraExprs = if (isNullAwareAntiJoin) h.expressions.head :: Nil else Nil
```

Comment:
Can we name extraExprs explicitly ? I am wondering if having this as an explicit different case branch (ie `case h : HashPartitioning if (isNullAwareAntiJoin)` would be in fact clearer ?